### ISSUE TITLE
Download OIPC Redline (Sign-Off and OIPC Review)

### ISSUE THREAD

**Issue by m-prodan** [bcgov/foi-flow#4596]
* As a user
* I want to be able to download the OIPC redline in PDF format
* so that I can give it (outside the system) to the ministry business area for sign off
* or the OIPC for review

**Assumptions & Scope**	
What are the assumptions for this story?

OIPC Redline for Ministry Sign-off will include all information unredacted, with redlines around the proposed redactions.

OIPC redline for OIPC review will include all information unredacted, EXCEPT for anything that is tagged for s. 14 - that information must be redacted.

Both OIPC Redline for Review and OIPC Redline for Ministry sign-off will NOT contain any pages flagged as Not Responsive of Duplicate

Ministry coordinator will not have access to the Record Review app; download of the redline will be triggered from the Records log.

IAO analysts can also download the redline records package.

The Records log filters are independent of the download. Applying a filter (e.g.by a divisional, error, etc. tag) on the Records Log will have no impact on what is downloaded.

When downloading a redline for Ministry sign-off you will get a ZIP file with one consolidated PDF for each Division. For example, if 4 divisions provided records there will be 4 redline PDFs in the ZIP file.

When downloading a redline for OIPC review, you will get a zip file with a single consolidated PDF

The ZIP file will also contain all the incompatible file types (i.e. non-PDFs), such as audio and GIS files.

What is IN scope?

Download the redline Records Package

What is NOT in scope?

**Acceptance Criteria**	

**Scenario 1: Active "Download Redline for Sign Off"**
* GIVEN the analyst has already created the OIPC Redline for Ministry Sign-Off
* WHEN the analyst clicks the Download dropdown
* THEN the Download dropdown will include the option: Download for Redline for Sign-Off
* AND the "Download Redline for Sign Off" will be active

**Scenario 2: "Download Redline for Sign Off" by Division**
* GIVEN the analyst has clicked "Download Redline for Sign Off" 
* WHEN the system completes creating the PDFs 
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/977423262
* THEN PDF(s) will be exported in one consolidated, linear PDF for each Division organized by modified date
* AND the consolidated PDF(s) redline for each Division will be include in the ZIP file
* AND the ZIP file will download onto the users computer
* AND the ZIP file will also include incompatible files in the correct divisional folders (i.e., the division that provided them)

**Scenario 3: Inactive "Download Redline for Sign Off"**
* GIVEN the analyst has NOT created the OIPC Redline for Ministry Sign-Off
* WHEN the coordinator or analyst is on the Download dropdown
* THEN the "Download Redline for Sign Off" will be inactive

**Scenario 4: Download Redline for Sign-Off - Error**
* GIVEN the analyst has saved the Redline
* WHEN the job fails at any point, including during the zipping service
* THEN the download redline for sign-off feature will be inactive
* AND the user will be notified that the redline for sign-off had an error and should be re-tried

**Scenario 5: Download Redline for Sign-Off - Only Incompatible Files**
* GIVEN the analyst has saved the Redline
* AND there is a division that only provided incompatible files, and will not have a redline
* WHEN the analyst downloads the Redline for Sign Off
* THEN the division that only provided incompatible files will have their own folder in the ZIP file
* AND it will contain all their incompatible files

**Scenario 6: Active "Download Redline for OIPC Review"**
* GIVEN the analyst has already created the OIPC Redline for OIPC Review
* WHEN the analyst clicks the Download dropdown
* THEN the Download dropdown will include the option: Download for Redline for OIPC review
* AND the "Download Redline for Sign Off" will be active

**Scenario 7: "Download Redline for OIPC Review**
* GIVEN the analyst has clicked "Download Redline for Sign Off" 
* WHEN the system completes creating the PDFs 
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/977423262
* THEN PDF will be exported in one consolidated, linear PDF by modified date
* AND the ZIP file will download onto the users computer

**Scenario 8: Inactive "Download Redline for OIPC Review"**
* GIVEN the analyst has NOT created the Redline for OIPC Review
* WHEN the coordinator or analyst is on the Download dropdown
* THEN the "Download Redline for OIPC Review" will be inactive

**Scenario 9: Download Redline for OIPC Review - Error**
* GIVEN the analyst has saved the Redline for OIPC Review
* WHEN the job fails at any point, including during the zipping service
* THEN the download redline for OIPC Review feature will be inactive
* AND the user will be notified that the redline for OIPC Review had an error and should be re-tried

**Scenario 10: Download Redline for OIPC Review - Incompatible Files**
* GIVEN the analyst has created the redline for OIPC review
* AND there are incompatible files in the records log
* WHEN the analyst downloads the Redline for OIPC Review
* THEN any incompatible files will also be included in the zip file alongside the single PDF

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Comment by KyEggleston** [comment]
@liseandtea @m-prodan - I think this needs to be reviewed to be consistent with #4170 and #4169 as well, as some of these ACs don't appear to be consistent with the new approach.

**Comment by divyav-aot** [comment]
**Refinement Comments:** 
1. New option in the records tab for Download Redlines for OIPC. This option will show only if the OIPC flag is there
2. Once it is ready to download, send notification to only IAO. Also, the Download options shouldn’t be visible for Ministry

**Comment by sumathi-thirumani-aot** [comment]
Need confirmation whether this should follow the concept of "Generate as single document" for all ministries or only for configured ministries. ?
@JHarrietha-AOT

**Comment by KyEggleston** [comment]
@sumathi-thirumani-aot 

When creating a redline for sign off (when the 'OIPC' review layer is created/selected) - captured under #4170 :

- User will select 'Create Response PDF' -> 'Redline for Sign Off' in the Redaction App
- Because the user has the 'OIPC' layer selected in the Redaction App, the redline will be produced based on that layer
- Because the 'Redline for Sign Off' button is being reused for this case, depending how that ministry is currently configured to receive redlines (either by division or as a consolidated PDF), they will continue to receive the redline in this manner. It does not matter if they produce a redline with the OIPC layer or Redline layer selected - they will always receive it based on how they are set up. 
- When complete, the user will go to the 'Records' tab. The user will select 'Download Redline for Sign Off', and they will receive a folder based on their current configuration (either by division or as a consolidated PDF). 

When creating a redline for OIPC review (when the 'OIPC' review layer is created/selected) - captured under #4169:

- User will select 'Create Response PDF' -> 'Redline for OIPC' in the Redaction App
- The 'Redline for OIPC' will be produced based on the OIPC layer, with all S.14 redacted (see ticket for more details)
- The 'Redline for OIPC' will always be produced as one, consolidated PDF. 
- When completed, the user will go to the 'Records' tab. The user will select 'Download Redline for OIPC Review', and they will receive a folder with one, consolidated PDF.

**Comment by KyEggleston** [comment]
I will clarify these ACs/scenarios, as they are a little confusing.

**Comment by KyEggleston** [comment]
@sumathi-thirumani-aot - discussed and agreed at standup adding hover over tooltip to display the last created date for the various packages listed under the 'Download' option in the Records tab. This would work similar to hovering over a Request # in Request Queue. This would allow analysts to see when various package were last created

![image](https://github.com/bcgov/foi-flow/assets/117312323/0970dc80-f36d-4251-8742-dbad335f97ee)

**Comment by Aman-Hundal** [comment]
@KyEggleston This has been completed  and will be pushed to test-rook soon!
